backend/de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/inout/ShipmentScheduleQtyReservedStorage.java

```diff
@@ -0,0 +1,134 @@
+import de.metas.material.cockpit.stock.StockDataQuery;
+import de.metas.material.event.commons.AttributesKey;
+import de.metas.product.ProductId;
+import de.metas.util.Services;
+import lombok.NonNull;
+import org.adempiere.inout.util.IShipmentScheduleQtyOnHandStorage;
+import org.adempiere.inout.util.ShipmentScheduleAvailableStockDetail;
+import org.adempiere.util.lang.impl.TableRecordReference;
+import org.adempiere.warehouse.WarehouseId;
+import org.adempiere.warehouse.api.IWarehouseDAO;
+import org.compiere.util.Util;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * This class shall hold details of reserved stock that's ready to be shipped (= VHUs that are reserved for one or more OrderLine).
+ */
+public class ShipmentScheduleQtyReservedStorage implements IShipmentScheduleQtyOnHandStorage
+{
+	private final CCache<Util.ArrayKey, StockDataQuery> cachedMaterialQueries = CCache.newLRUCache("QtyReservedCachedMaterialQueries", 200, CCache.EXPIREMINUTES_Never);
+	private final IShipmentScheduleEffectiveBL shipmentScheduleEffectiveBL = Services.get(IShipmentScheduleEffectiveBL.class);
+	private final IWarehouseDAO warehouseDAO = Services.get(IWarehouseDAO.class);
+	@NonNull final ImmutableList<ShipmentScheduleAvailableStockDetail> stockDetails;
+
+	public ShipmentScheduleQtyReservedStorage(final @NonNull ImmutableList<ShipmentScheduleAvailableStockDetail> stockDetails)
+	{
+		this.stockDetails = stockDetails;
+	}
+
+	@Override
+	public List<ShipmentScheduleAvailableStockDetail> getStockDetailsMatching(final @NonNull I_M_ShipmentSchedule sched)
+	{
+		if (stockDetails.isEmpty())
+		{
+			return Collections.emptyList();
+		}
+		else
+		{
+			//
+			// Reserved qty shall always be in VHUs, no need for picking BOM.
+			final StockDataQuery mainProductQuery = toQuery(sched);
+			return new ArrayList<>(getStockDetailsMatching(mainProductQuery));
+		}
+	}
+
+	private ImmutableList<ShipmentScheduleAvailableStockDetail> getStockDetailsMatching(@NonNull final StockDataQuery query)
+	{
+		return stockDetails
+				.stream()
+				.filter(stockDetail -> matching(query, stockDetail))
+				.collect(ImmutableList.toImmutableList());
+	}
+
+	private static boolean matching(final StockDataQuery query, final ShipmentScheduleAvailableStockDetail stockDetail)
+	{
+		//
+		// Product
+		if (!ProductId.equals(query.getProductId(), stockDetail.getProductId()))
+		{
+			return false;
+		}
+
+		//
+		// Warehouse
+		final Set<WarehouseId> queryWarehouseIds = query.getWarehouseIds();
+		if (!queryWarehouseIds.isEmpty() && !queryWarehouseIds.contains(stockDetail.getWarehouseId()))
+		{
+			return false;
+		}
+
+		//
+		// Attributes
+		final boolean queryMatchesAll = query.getStorageAttributesKey().isAll();
+		final boolean queryMatchesStockDetail = AttributesKey.equals(query.getStorageAttributesKey(), stockDetail.getStorageAttributesKey());
+		return queryMatchesAll || queryMatchesStockDetail;
```

Comment:
afaiu, atm `query.getStorageAttributesKey()` will always be `ALL`, is it still WIP?